### PR TITLE
Add egs_dynamic_geometry and egs_view time viewing

### PR DESCRIPTION
**Pull request by a-l-demelo**
Add the ability for dynamic geometries, including translation and rotation of any geometry. This is similar to `egs_dynamic_source`. Time indices are synchronized with any source that includes time indexes such as `egs_beam_source`, `IAEAphsp_source`, and `egs_dynamic_source`. 

Add a time slider to `egs_view` that shows geometry motion. Includes a time window for showing tracks over a window of time, and the tracks update as you move the time index slider to show tracks synchronized with geometry motion.

### CODE REVIEW

**Review by rtownson — APPROVED** [review]

**Comment by blakewalters** [comment]
This is excellent!  I tested out dynamic motion of a fairly complex rotating geometry within a stationary one (I also threw in some translation for good measure), and I was able to display the track evolution in egs_view beautifully.  I would just clarify the documentation at the top of the egs_dynamic_geometry.h a bit by:
1) changing the notation for the translation from "x(N) y(N) z(N)" to "xtransl(N) ytransl(N) ztransl(N)"
2) indicating that rotations are defined in degrees and translations in cm.
Otherwise, let's get this merged forthwith!

File: HEN_HOUSE/egs++/egs_transformations.cpp [hunk 1]
```diff
@@ -67,7 +68,7 @@ EGS_AffineTransform *EGS_AffineTransform::getTransformation(EGS_Input *i) {
     if (!err && tmp.size() == 3)
         result = new EGS_AffineTransform(
             EGS_RotationMatrix(EGS_Vector(tmp[0],tmp[1],tmp[2])),t);
-    else {
+    else {       //egsWarning("getTransformation: rotation specified by 3 angles: %g   %g   %g\n\n",tmp[0],tmp[1],tmp[2]);
```
Comment (ftessier) [inline HEN_HOUSE/egs++/egs_transformations.cpp:71]: Remove spurious debug output that is commented out.

File: HEN_HOUSE/egs++/egs_transformations.cpp [hunk 2]
```diff
@@ -105,3 +106,32 @@ EGS_AffineTransform *EGS_AffineTransform::getTransformation(EGS_Input *i) {
     }
     return result;
 }
+
+//////////// OVERLOADING FOR DYNAMIC GEOMETRY /////////////
+EGS_AffineTransform *EGS_AffineTransform::getTransformation(vector<EGS_Float> translation, vector<EGS_Float> rotation) {
+    /* The following method is used by the dynamic geometry class to create a transformation corresponding to a certain translation and rotation vector. These vectors are determined
+     * through the control points and the sampled mu value. The sampled transformation coordinates are passed here to build a transformation */
```
Comment (ftessier) [inline HEN_HOUSE/egs++/egs_transformations.cpp:113]: Change `mu value` to `time value`

File: HEN_HOUSE/egs++/view/image_window.cpp [hunk 3]
```diff
@@ -664,4 +667,4 @@ void ImageWindow::handleAbort() {
             renderState = WorkerIdle;
         }
     }
-}
+}
```
Comment (ftessier) [inline HEN_HOUSE/egs++/view/image_window.cpp:670]: Add newline at end of file

File: HEN_HOUSE/egs++/view/viewcontrol.cpp [hunk 4]
```diff
@@ -62,6 +65,10 @@
 #include <vector>
 #include <algorithm>
 #include <unordered_map>
+#include<unistd.h>
```
Comment (ftessier) [inline HEN_HOUSE/egs++/view/viewcontrol.cpp:68]: Add space after `#include`

File: HEN_HOUSE/egs++/view/viewcontrol.cpp [hunk 5]
```diff
@@ -249,6 +254,7 @@ bool GeometryViewControl::loadInput(bool reloading, EGS_BaseGeometry *simGeom) {
             g = 0;
         }
         EGS_BaseGeometry::clearGeometries();
+        geometryNames.clear(); // adding this to solve loading new input file from file tab in egsview. Previously name array is never cleared after loading new input file
```
Comment (ftessier) [inline HEN_HOUSE/egs++/view/viewcontrol.cpp:257]: 🎉

File: HEN_HOUSE/egs++/geometry/egs_dynamic_geometry/egs_dynamic_geometry.cpp [hunk 6]
```diff
@@ -0,0 +1,601 @@
...
+#  WARRANTY; without even the implied warranty of MERCHANTABILITY or FITNESS
+#  FOR A PARTICULAR PURPOSE.  See the GNU Affero General Public License for
+#  more details.
+#
+#  You should have received a copy of the GNU Affero General Public License
+#  along with EGSnrc. If not, see <http://www.gnu.org/licenses/>.
+#
+###############################################################################
+#
+#  Author:          Alex Demelo, 2023
```
Comment (ftessier) [inline HEN_HOUSE/egs++/geometry/egs_dynamic_geometry/egs_dynamic_geometry.cpp:24]: Should be `Alexandre Demelo`, for consistency with other files.

File: HEN_HOUSE/egs++/geometry/egs_dynamic_geometry/egs_dynamic_geometry.h [hunk 7]
```diff
@@ -0,0 +1,433 @@
...
+#  WARRANTY; without even the implied warranty of MERCHANTABILITY or FITNESS
+#  FOR A PARTICULAR PURPOSE.  See the GNU Affero General Public License for
+#  more details.
+#
+#  You should have received a copy of the GNU Affero General Public License
+#  along with EGSnrc. If not, see <http://www.gnu.org/licenses/>.
+#
+###############################################################################
+#
+#  Author:          Alex Demelo 2023
```
Comment (ftessier) [inline HEN_HOUSE/egs++/geometry/egs_dynamic_geometry/egs_dynamic_geometry.h:24]: Should be `Alexandre Demelo`, for consistency with other files.

**Review by ftessier — CHANGES_REQUESTED** [review]
### Synchronized .ptracks file

I wonder why a `.syncptracks` extension is added to distinguish static and dynamic tracks files. I think we ought to avoid adding extensions where possible. I would keep `.ptracks` for all tracks file, and simply add a `0` time to files that do not track time. If the purpose of a new extension is to curb file size or address backward compatility, then this ought to be handled by a flag _inside_ the file (or otherwise determined from the file content), not in the file extension!

### Time instead of monitor units

The language (text, variables, functions) were changed from `mu` to `time`, which is great for clarity (`mu` lingo is specific to medical physics). However, there subsists a few instances of `mu` in `egs_phsp_scoring`. If that is not required for some other reason, I suggest updating those as well:

```c
HEN_HOUSE/egs++/ausgab_objects/egs_phsp_scoring/egs_phsp_scoring.h
80:IAEA format: iq,E,[x],[y],[z],u,v,wt,latch,[mu]
85:user has the option of scoring the synchronization parameter, mu, passed from the source.
100:          score mu                 = yes or no [default] (IAEA format only)
127:In IAEA-format phase space files, the user can opt to score the synchronization parameter, mu.
131:synchronized CMs or scored using this ausgab object with mu scoring turned on) and
215:        score mu = no
451:        EGS_Float E, x, y, z, u, v, w, wt, mu;
493:    int mu_ind; //IAEA id for mu--no ID for now so set to generic float (0)
501:    bool score_mu; //set to true if scoring mu
502:    float pmu; //mu value associated with particle

HEN_HOUSE/egs++/ausgab_objects/egs_phsp_scoring/egs_phsp_scoring.cpp
159:            iaea_i_mu=0; //position of mu in array
188:        description += "\n mu will be scored (if available)";
244:    //if user requested mu scoring, check if mu is available
246:        egsWarning("\nEGS_PhspScoring: User requested mu scoring, but mu is inavailable with this source.\n");
247:        egsWarning("Turning off mu scoring.\n");
281:        p_stack[phsp_index].mu = app->getTimeIndex();
406:                iaea_extra_float[iaea_i_mu] = p_stack[j].mu;
633:                //see if user wants to score mu (if available)
635:                if (!input->getInput("score mu", str)) {
639:                    imuscore = input->getInput("score mu",allowed_muscore,-1);
641:                        egsWarning("\nEGS_PhspScoring: Invalid input for mu scoring.  Will not score mu.\n");
```

### Minor comments

- [x] Resolve all conversations left as individual comments.

**Comment by ftessier** [comment]
@a-l-demelo if you have please comment on the review comments. In the meantime I will address the minor issues.

File: HEN_HOUSE/egs++/egs_base_geometry.h [hunk 8]
```diff
@@ -149,6 +151,16 @@ class EGS_EXPORT EGS_BaseGeometry {
     */
     virtual int isWhere(const EGS_Vector &x) = 0;
 
+    /* getNextGeom is the equivalent of getNextParticle but for the simulation object. It's goal is to determine the next state of the geometry, either by synchronizing itself to the
+     * source time parameter, or by sampling it's own time parameter and updating itself accordingly if the source has provided no time index.
+     *
+     *This function has a non-empty implementation in 2 cases.
+     *1) it is re implemented in any composite geometry, where it will call next geom on all of its components
```
[the remark this answers is not in the document]
Comment (ftessier) [reply HEN_HOUSE/egs++/egs_base_geometry.h:158]: Should be `reimplemented` (no hyphen, see [discussion](https://english.stackexchange.com/questions/310040/reimplement-or-re-implement)).

**Review by blakewalters — APPROVED** [review]

**Comment by rtownson** [comment]
Fred's suggestions have been resolved.

**Review by mainegra — APPROVED** [review]

**Comment by rtownson** [comment]
Closing this because it's superseded by #1197.

**Review by ftessier — APPROVED** [review]

**Pull request closed without merging**